jdaviz/configs/cubeviz/plugins/parsers.py

```diff
@@ -68,49 +71,34 @@ def _parse_hdu(app, hdulist, file_name=None):
 
     file_name = file_name or "Unknown HDU object"
 
-    # WCS may only exist in a single extension (in this case, just the flux
-    #  flux extention), so first find and store then wcs information.
-    wcs = None
-
-    for hdu in hdulist:
-        if hdu.data is None or not hdu.is_image:
-            continue
-
-        try:
-            sc = SpectralCube.read(hdu, format='fits')
-        except (ValueError, FITSReadError):
-            continue
-        else:
-            wcs = sc.wcs
-
     # Now loop through and attempt to parse the fits extensions as spectral
     #  cube object. If the wcs fails to parse in any case, use the wcs
     #  information we scraped above.
     for hdu in hdulist:
         data_label = f"{file_name}[{hdu.name}]"
 
-        if hdu.data is None or not hdu.is_image:
+        if hdu.data is None or not hdu.is_image or len(hdu.data.shape) != 3:
             continue
 
-        # This is supposed to fail on attempting to load anything that
-        # isn't cube-shaped. But it's not terribly reliable
+        wcs = WCS(hdu.header)
+
+        if 'BUNIT' in hdu.header:
+            flux = hdu.data * u.Unit(hdu.header['BUNIT'])
+        else:
+            # TODO Add warning that actual units were not found and using fake ones
+            logging.warn("No flux units found in hdu, using Jansky as a stand-in")
+            flux = hdu.data * u.Jy
```

Comment:
I agree with @duytnguyendtn because `u.Unit(hdu.header['BUNIT'])` is going to break for most of the FITS images you have. For example: `'ELECTRONS/S'` does not resolve to `u.electron/u.s` (`astropy.units` is case-sensitive and very pedantic about the "s"es).

However, falling back to `u.Jy` is also a problem. For example, when parsing of `'ELECTRONS/S'` fails, you get Jy, which is wrong.

Reply:
To start with the axes, moving the axes using `np.moveaxis(flux, 1, 2)` or `wcs = wcs.swapaxes(0, 1)` does NOT change the x or y axis labels in the glue jupyter viewer. Those labels are set using the `ComponentIDs`, I believe in glue astronomy, but hours investigating this did yield useful results. We may need some SME help on this matter.

In regards to the flux units, thank you for your suggestions Duy! I will implement your suggestions in the next commit. To answer Pey Lian, I agree that giving the user other units is a problem but I think the choice is either that or not load data at all. Although, we were having discussions previously about using `counts` as units, is that a better replacement?

Reply:
I think this may be a symptom of the broader issue @aaastorrs brought up in #566. I'd argue you're right, @javerbukh; I don't think we shouldn't have to rely on assigning an arbitrary `u.Jy` when things don't work (as @pllim brings up, this can very often create a graph that is lying to the user), rather I'd argue that a generic `u.count` is at least a "correct" fallback since it is so generic if I understand the unit right

Reply:
Yes, we use "count" for anything here. 😆

Reply:
![DzhyYE8d_400x400](https://user-images.githubusercontent.com/2090236/116892639-06db2480-abfe-11eb-9aa6-4924a0185f47.jpg)